### ISSUE TITLE
Subclass of an abstract class is not instantiable with type narrowing

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#3543]
## Environment data

-   Language Server version: 2021.5.1
-   OS and version: Windows 7
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.12 Anaconda

## Code Snippet

```python
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import cast

class Base(ABC):
    @abstractmethod
    def f(self) -> None:
        ...

cls = globals()["..."]  # Fetch a value with type `Any`.
assert issubclass(cls, Base)

if False:
    cls = cast("type[Base]", cls)

_ = cls()
```

## Expected behavior

`cls` is treated as instantiable.

## Actual behavior

`cls` is abstract and not instantiable.

If the `cast` line got enabled, `cls` would become instantiable, though its type would seem unchanged.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.253, which I just published. It will also be included in the next release of pylance.

**Comment by Azureblade3808** [comment]
```python
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import cast

class Base(ABC):
    @abstractmethod
    def f(self) -> None:
        ...

cls = globals()["..."]  # Fetch a value with type `Any`.
assert isinstance(cls, type)  # <-- Added this line.
assert issubclass(cls, Base)

if False:
    cls = cast("type[Base]", cls)

_ = cls()
```

Added a sanity check of being instance of `type` and got warnings again.

In real code, these are checks and returns instead of assertions, and `instanceof(..., type)` must be run before `issubclass(..., ...)`, otherwise the code may raise exceptions when the value is not a class.
